numba/targets/arraymath.py

```diff
@@ -3666,6 +3666,48 @@ def _is_nonelike(ty):
     return (ty is None) or isinstance(ty, types.NoneType)
 
 
+@overload(np.array)
+def np_array(a, dtype=None, copy=True):
+
+    # developer note... keep this function (type_can_asarray) in sync with the
+    # accepted types implementations below!
+    if not type_can_asarray(a):
+        return None
+
```

Comment:
This needs to deal with optional values where `dtype` could be a `value` or `None`. Example:
```python
from numba import njit
import numpy as np

@njit
def foo(z):
    dt = None
    if z > 2:
        dt = np.float32
    # dtype here could be None or could be np.float32, won't know until runtime
    # this is encoded in a `types.Optional(numba.typeof(np.float32))`
    return np.array((1,2,3), dtype=dt)

x = foo(1)
print(x)
print(x.dtype)
```
I think if the `dtype` is an `Optional` then it's impossible to determine statically what should happen in all cases apart from `dtype=None` or `dtype="type of input array"`, because the `value` or  `None` will influence the type of the returned item.